### PR TITLE
[MLIR][Arith] Add ExpandOps to convertArithToLLVM

### PR DESCRIPTION
**Pull request by nujaa**
Arith Floor and Ceil ops would not get lowered when running --convert-arith-to-llvm.

### CODE REVIEW

**Comment by banach-space** [comment]
Hey Hugo, thanks for sending this! This is marked as draft, so I presume not yet ready for reviewing?

**Comment by nujaa** [comment]
> Hey Hugo, thanks for sending this! This is marked as draft, so I presume not yet ready for reviewing?

Sorry I forgot to submit it after running the draft tests. 🙄

**Review by krzysz00 — CHANGES_REQUESTED** [review]
Given the context, I'll argue for `mlir::arith::populateCeilFloorDivExpandOpsPatterns` specifically.

The other "expand" patterns include things like expanding `minsi` or `maxui`, which have LLVM lowerings, and we don't want to add those here.

File: mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp
Comment (joker-eph) [inline mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp:511]: This pass is kind of deprecated: we're supposed to migrate to `--convert-to-llvm`. 
I'm concerned by additions here that wouldn't carry over to `--convert-to-llvm`!

Comment (nujaa) [reply mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp:511]: I am not completely sure of what you mean. 
The tests `mlir/test/Conversion/ArithToLLVM/arith-to-llvm.mlir` calls both `convert-arith-to-llvm` and `--convert-to-llvm`. If either arith-to-llvm or convert-to-llvm is updated, the other must be updated as well.

**Comment by nujaa** [comment]
> Given the context, I'll argue for `mlir::arith::populateCeilFloorDivExpandOpsPatterns` specifically.
> 
> The other "expand" patterns include things like expanding `minsi` or `maxui`, which have LLVM lowerings, and we don't want to add those here.

Good catch, thanks.

**Review by krzysz00 — APPROVED** [review]
Removing my hold

File: mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp
Comment (banach-space) [reply mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp:511]: @nujaa , wouldn't adding this in "ConvertToLLVMPass.cpp" _instead_ also work?

Comment (joker-eph) [reply mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp:511]: I'm confused because you're only changing the `ArithToLLVMConversionPass` here, so was this behavior already implemented in the `convert-to-llvm` flow and you're just catching up?

(I guess that makes my point as well somehow: we should remove the `convert-arith-to-llvm` pass: it does not provide anything that `convert-to-llvm` can't do, hopefully).

Comment (nujaa) [reply mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp:511]: > @nujaa , wouldn't adding this in "ConvertToLLVMPass.cpp" _instead_ also work?

I think it is Arith specific and should not be part of generic ConvertToLLVMPass

Comment (nujaa) [reply mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp:511]: > so was this behavior already implemented in the `convert-to-llvm` flow and you're just catching up?

Neither arith-to-llvm nor convert-to-llvm was not expanding ops.

**Comment by RoboTux** [comment]
Another option is to move that code that into MLIRArithUtils if we do not want to depend on MLIRArithTransforms

**Comment by banach-space** [comment]
Reverted.

@nujaa , this shouldn't land while there's an ongoing design discussion on the suitable placement of this patterns.

**Comment by nujaa** [comment]
> @nujaa , this shouldn't land while there's an ongoing design discussion on the suitable placement of these patterns.

Yes, Sorry, I didn't see the upper comments until I merged following the approval and t. My bad.

**Comment by nujaa** [comment]
Alright. Thanks both for reverting, pointing it out and the motivation. I could not find guidelines on how to revert a commit without going through the MR review process. Do I simply push the revert commit without waiting for approval ?

Before I submit any fix, I would like to make sure of a few things : 
To answer @joker-eph ,
>I'm confused because you're only changing the `ArithToLLVMConversionPass`

If I am not mistaken, `convert-to-llvm` calls `populateConvertToLLVMConversionPatterns` implemented by a `ConvertToLLVMPatternInterface`  extended for Arith behind the `ArithToLLVMDialectInterface`. This patch does adds a call to populateCeilFloorDivExpandOpsPatterns to both `ArithToLLVMConversionPass` (the deprecated pass) AND `ArithToLLVMDialectInterface::populateConvertToLLVMConversionPatterns` (the preferred one). I am not sure if I should do something else to mention it is part of convert-to-llvm.

> wouldn't adding this in "ConvertToLLVMPass.cpp" instead also work?

I think `ConvertToLLVMPass.cpp` implements a shared interface among dialects to lower to LLVM and should not implement a requirement for some ops of a given dialect. 
@banach-space , do you maybe mean adding CeilFloorDivExpandOpsPatterns to `populateArithToLLVMConversionPatterns` which is called in both `ArithToLLVMConversionPass` AND `ArithToLLVMDialectInterface::populateConvertToLLVMConversionPatterns` ? I agree it does factorize the code. Although, `populateArithToLLVMConversionPatterns` calls specifically lowerings of arith ops to LLVM of the type `%OpName%Lowering`. Whereas CeilFloorDivExpandOpsPatterns is not an op lowering to LLVM per se. I am convincible on this, you have the last word.

Thanks again all for your patience.

**Comment by banach-space** [comment]
>  I could not find guidelines on how to revert a commit without going through the MR review process.

There's a few relevant parts to consider here. First, the revert policy:
*  https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy

And indeed, it does not discuss "how". However, note:
> Reverts should be reasonably timely. A change submitted two hours ago can be reverted without prior discussion.

And then, under [obtaining-commit-access](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access)
> You are allowed to commit patches without approval which you think are obvious. This is clearly a subjective decision — we simply expect you to use good judgement.

From my experience, "reverts" are where you tend to 100% rely on your judgement - it's the easiest and quickest way to fix buildbot failures. In more complex situation (e.g. no buildbot breakage upstream, but some issues downstream), we tend use PRs. But even then, if a repro has been shared, folks tend to revert. This way patch authors are _not under pressure_ to fix immediately (sometimes that's just physically not possible).

> Thanks again all for your patience.

No worries Hugo, this is standard practice ;-)

**Comment by banach-space** [comment]
> If I am not mistaken, convert-to-llvm calls populateConvertToLLVMConversionPatterns implemented by a ConvertToLLVMPatternInterface extended for Arith behind the ArithToLLVMDialectInterface. 

OK, this is what we were missing :) And apologies, I forgot how `-convert-to-llvm` was wired-up. 

Like you said, `ConvertToLLVM` invokes these `populateConvertToLLVMConversionPatterns` methods for every Dialect that implements [ConvertToLLVMPatternInterface](https://github.com/llvm/llvm-project/blob/52b9d0beb6225d1ffe89e2921774f0df0b3e7969/mlir/include/mlir/Conversion/ConvertToLLVM/ToLLVMInterface.h#L24-L46):
https://github.com/llvm/llvm-project/blob/52b9d0beb6225d1ffe89e2921774f0df0b3e7969/mlir/lib/Conversion/ConvertToLLVM/ConvertToLLVMPass.cpp#L118-L122

And that's what's being updated here. But you are also updating the `-convert-arith-to-llvm` path, right? In summary:
* By updating [ArithToLLVMDialectInterface](https://github.com/llvm/llvm-project/blob/52b9d0beb6225d1ffe89e2921774f0df0b3e7969/mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp#L522-L535), you are updating `-convert-to-llvm`. 
* By updating [ArithToLLVMConversionPass](https://github.com/llvm/llvm-project/blob/52b9d0beb6225d1ffe89e2921774f0df0b3e7969/mlir/lib/Conversion/ArithToLLVM/ArithToLLVM.cpp#L494-L513), you keep `-convert-artih-to-llvm` in sync with `-convert-to-llvm`?

Is this accurate?

> Arith Floor and Ceil ops would not get lowered when running --convert-arith-to-llvm.

Considering Mehdi's comment re this path being deprecated in favour of `-convert-to-llvm`, shouldn't this comment focus on `-convert-to-llvm` instead? Also, some element of this discussion should be included in the summary. We shouldn't be required to scan PR discussions to understand the rationale behind a particular change - that's what summaries are for ;-)

**Comment by joker-eph** [comment]
The confusion to me came from comments like:

> I think it is Arith specific and should not be part of generic ConvertToLLVMPass

I think an answer that would have resolved the discussion immediately could have been along the line of: 

> this patch is actually plumbing the new pattern through both passes, since this is adding the call in `ArithToLLVMDialectInterface` as well. The test demonstrates the behavior is consistent between the two passes.

We should be all good now I think.

**Pull request merged**
